src/collections/unordered-map.ts

```diff
@@ -59,21 +59,21 @@ export class UnorderedMap {
     return keysIsEmpty;
   }
 
-  get(key: Bytes): unknown | null {
+  get(key: Bytes): T | null {
     let indexRaw = getIndexRaw(this.keyIndexPrefix, key);
     if (indexRaw) {
       let index = deserializeIndex(indexRaw);
       let value = this.values.get(index);
       if (value) {
-        return value;
+        return value as T;
```

Comment:
`as T` also doesn't solve the problem, it's only a ts sytax suger. When inspect the built JS contract, it is ignored. For example: https://www.typescriptlang.org/play?#code/LAKFGMBsEMGdYAQGFoCcDKAHApuBBvUBYhcAewDtYAXVAV3GrNQApM6AjSASzwugC22AFwIaqbhQDmASgIBfUIrAhI2aggAeCALwF+Q0QHJoR5Wo0BPXVoRxkaLLgDcESrDJqAdJDJSW1pI00BTg2GQAZg4YOOAyQA

Reply:
@ailisp This could be solved by the new _brand_ feature of ES2022 ([here](https://inspiredwebdev.com/everything-new-in-es2022/#ergonomic-brand-checks-for-private-fields)). Using TS syntax here is better for DX since the developer won't look at the compiled code anyway.

Reply:
Looks like this feature is not related to deserialization. Can you please elaborate on how it can help?

Reply:
When constructing a collection we could set the private field (_brand_) to the type of the underlying data we store, this would be serialized and stored which means we could theoretically infer the type with this field when deserializing the collection. We could also just let the user decide what the type is which is easier, and I don't know how valuable inferring the type is anyway. That's why I am saying it doesn't really matter that it says `as T` since types in JS/TS are just for developer experience anyway

Reply:
I'm not sure that this solution here is 100% okay, but something in this direction could go: 

https://www.typescriptlang.org/play?#code/LAKFGMBsEMGdYAQGFoCcDKAHApuBBvUBYhcAewDtYAXVAV3GrNQApM6AjSASzwugC22AFwIaqbhQDmASgIBfUIrAqocRAFUKzACbZU2HQFlomADwAVAHwEiJclVoMmrdl14ID0HZUgBPBEkaaApwEQQLOUIQEgRlO2IpbGoWGVELBAAfBAo6SEhbGNiSSGSEAA8EAF4CfiFRAHJoBuVikgB6doABalgAWm4pbQME4oNqOlQKHOwAdwRqAAtuWAA6IOoQsJZymVHY0eV4kFLqBAFTapn5rV19QxNMFhQMHHAZAG4IShoEAJqLphVkkUnsQA5YGRSqtIGQpCwAhsttgyAAzZBoLC4AA0fwA-Ks6tgwUA

I don't know if it's okay to ask the user to pass us an instance of the class to constructor :D

Reply:
Allowing partial deserialization (simple `as DataType`) is also an option. Maybe JS developers are used to this behavior and do not expect functions to be in place.

Reply:
@volovyks @osalkanovic Speaking as a JS dev, this behavior is totally fine. One convenience that could be added to something like this, because of the nature of serializing classes and collections could be something like the following:
```typescript
get(key: Bytes, deserializer?: (value: unknown) => T): T | null {
  const indexRaw = getIndexRaw(this.keyIndexPrefix, key);

  if (indexRaw) {
    const index = deserializeIndex(indexRaw);
    const value = this.values.get(index);
    
    if (value) {
      return !!deserializer ? deserializer(value) : (value as T);
    } else {
      throw new Error(ERR_INCOSISTENT_STATE);
    }
  }

  return null;
}
```

Reply:
@petarvujovic98  Nice, I agree. 
If @volovyks  and @ailisp  agree, I would correct and update the PR.

Reply:
@petarvujovic98 what are you think about this [example](https://github.com/near/near-sdk-js/pull/203#discussion_r964503394)?

Reply:
@osalkanovic I wouldn't use this approach as the instance is going to also get serialized so you will lose the constructor, plus if it is a primitive type it won't need to have a constructor and you are complicating the DX. 

Keeping track of type data after seralization is difficult in JS, but since users are using TS to write the construction and access of variables after deserialization, it's better to keep the logic for those in the TS layer rather than in the JS/runtime layers.

Reply:
I agree with @petarvujovic98 , and the way @osalkanovic shown is quite clever that I didn't think of.  May @osalkanovic 's way be incorporated in as a pattern to write `deserializer?: (value: unknown) => T`

Reply:
@ailisp I will update PR with deserializer today.

Reply:
@ailisp @volovyks @petarvujovic98 PR is ready for review